nova/virt/vmwareapi/vmops.py

```diff
@@ -3172,15 +3176,60 @@ def _sync_sync_server_group(context, sg_uuid):
                 LOG.debug('Sync for server-group %s done', sg_uuid)
                 return
 
+            # First we check for all instances, which have ongoing migrations
+            # on the given host, either as source or destination
+
+            # Decision matrix for migrations:
+            # Mig-Status Action/Host
+            #            Source  Dest
+            # Preparing  Remove  N/A
+            # Running    Remove  Add
+            # (Other states are consistent with default behaviour)
+            #
+            # So the Instance.host will always be the source of the migration,
+            # and we want to remove the rules.
+            # We only need to handle specially the case a running migration
+            # on the destination host, and add it
+
+            MigrationList = objects.migration.MigrationList
+            filters = {
+                "host": host,
+                "instance_uuid": sg.members,
+                "status": ["preparing", "running"],
+            }
+
+            expected_members = {}
+
+            migrations_by_instance_uuid = {}
+            for migration in MigrationList.get_by_filters(context, filters):
+                instance_uuid = migration.instance_uuid
+                if migration.source_compute == host:
+                    # The host is the source of a migration
+                    # That means the instance will be part of the instance list
+                    # So we have to remember that instance to be removed from
+                    # the DRS rule-set
+                    migrations_by_instance_uuid[instance_uuid] = \
+                        migration
+                else:
+                    # We now handle the destination side
+                    if migration.state == "preparing":
+                        # Not even started, we can ignore that one
+                        continue
+
+                    # Polling the cache, as vm_util.get_vm_ref is very slow
+                    # for the negative search.
```

Comment:
Can we fill the cache in the live-migration before calling `sync_server_group()`?

Reply:
That is the idea. If the VM is on the destination host, we set the cache value.
We need to do that to work with the instance in any way anyway.